server/src/main/kotlin/suwayomi/tachidesk/global/impl/sync/SyncManager.kt

```diff
@@ -0,0 +1,379 @@
+            .apply()
+    }
+
+    private fun getAllMangaFromDB(): List<MangaDataClass> {
+        return transaction { MangaTable.selectAll().map { MangaTable.toDataClass(it) } }
+    }
+
+    private fun getAllMangaThatNeedsSync(): List<MangaDataClass> {
+        return transaction {
+            MangaTable.selectAll().where { MangaTable.inLibrary eq true }.map { MangaTable.toDataClass(it) }
+        }
+    }
+
+    private fun isMangaDifferent(localManga: MangaDataClass, remoteManga: BackupManga): Boolean {
+        val localChapters = transaction {
+            ChapterTable
+                .selectAll()
+                .where { ChapterTable.manga eq localManga.id }
+                .map { ChapterTable.toDataClass(it) }
+        }
+        val localCategories = transaction {
+            CategoryMangaTable
+                .innerJoin(CategoryTable)
+                .selectAll()
+                .where { CategoryMangaTable.manga eq localManga.id }
+                .map { it[CategoryTable.order] }
+        }
+
+        if (areChaptersDifferent(localChapters, remoteManga.chapters)) {
+            return true
+        }
+
+        if (localManga.version != remoteManga.version) {
+            return true
+        }
+
+        if (localCategories.toSet() != remoteManga.categories.toSet()) {
+            return true
+        }
+
+        return false
+    }
+
+    private fun areChaptersDifferent(
+        localChapters: List<ChapterDataClass>,
+        remoteChapters: List<BackupChapter>,
+    ): Boolean {
+        val localChapterMap = localChapters.associateBy { it.url }
+        val remoteChapterMap = remoteChapters.associateBy { it.url }
+
+        if (localChapterMap.size != remoteChapterMap.size) {
+            return true
+        }
+
+        for ((url, localChapter) in localChapterMap) {
+            val remoteChapter = remoteChapterMap[url]
+
+            // If a matching remote chapter doesn't exist, or the version numbers are different, consider them different
+            if (remoteChapter == null || localChapter.version != remoteChapter.version) {
+                return true
+            }
+        }
+
+        return false
+    }
+
+    private fun filterFavoritesAndNonFavorites(backup: Backup): Pair<List<BackupManga>, List<BackupManga>> {
+        val favorites = mutableListOf<BackupManga>()
+        val nonFavorites = mutableListOf<BackupManga>()
+
+        val elapsedTimeMillis = measureTimeMillis {
+            val databaseManga = getAllMangaFromDB()
+            val localMangaMap = databaseManga.associateBy {
+                Triple(it.sourceId.toLong(), it.url, it.title)
+            }
+
+            logger.debug { "Starting to filter favorites and non-favorites from backup data." }
+
+            backup.backupManga.forEach { remoteManga ->
+                val compositeKey = Triple(remoteManga.source, remoteManga.url, remoteManga.title)
+                val localManga = localMangaMap[compositeKey]
```

Comment:
Make this just run a bunch of db queries instead of loading all the manga from the db into memory. It might not be faster but it would be better.